### PR TITLE
[WIP] Try to fix dist-upgrade

### PR DESCRIPTION
**Pull request by fauust**
Let's see if BB pass.
Discussion is in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=988089 for the moment.

### CODE REVIEW

**Comment by ottok** [comment]
Galera used to be a 'recommends' but was changed to 'depends' in 2015: https://github.com/MariaDB/server/commit/b9fee60871264ea56a5582ca7a436aabdabea6fc

See also Zulip discussion: https://mariadb.zulipchat.com/#narrow/stream/118759-general/topic/Debian/near/239011054

If/when Galera is made a 'recommends' again, please also make the same change for RPM packages so that both packagings have the same logic for end users.

..and this can certainly not be done on the 10.5 branch, but only in a new major release (e.g. 10.7.)

**Comment by grooverdan** [comment]
FYI galera was always a requires on RPM since it was introduced https://github.com/MariaDB/server/commit/df4dd593f29aec8e2116aec1775ad4b8833d8c93#diff-b34a4a2b95c6341e79aef8cb1043e66dd672ee00df1382a13582c04e99c08bee

**Comment by vuvova** [comment]
I would also think galera should be recommended. I totally agree with the original comment in 6bd94cf54274d54521ece1e50d534777122ff29e

**Review by SergMariaDB — APPROVED** [review]

**Review by ottok — CHANGES_REQUESTED** [review]
Come on, you can't have a commit with just a subject and that subject being "Try to fix dist-upgrade".. 
See https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project for tips on the topic.

Also:
- this PR must target 10.7, this change cannot go into a stable release
- this was already done in 6bd94cf, redoing it now should be quite easy, right?
- the equivalent change should also go into RPM packaging (similar to how #1556 is pending)

When this PR is ready, please remove "WIP" from title and request re-review. Thanks!

**Comment by fauust** [comment]
> Come on, you can't have a commit with just a subject and that subject being "Try to fix dist-upgrade".. See https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project for tips on the topic.

Come on @ottok, do you really think that this commit was destined to be definitive (I thought that the PR subject and comment was pretty clear on what my first intention was) :smile: Sorry for letting this PR so much time in `WIP` though.

> * this PR must target 10.7, this change cannot go into a stable release

Yep

> * this was already done in [6bd94cf](https://github.com/MariaDB/server/commit/6bd94cf54274d54521ece1e50d534777122ff29e), redoing it now should be quite easy, right?

This is not exactly the same change, but as @vuvova also confirmed that 6bd94cf is the way to go, I will.

> * the equivalent change should also go into RPM packaging (similar to how #1556 is pending)

@illuusio, could you please help me on that one?

**Comment by illuusio** [comment]
TL;DR problem. Just please tell me why is Galdera hard dependency. I kind of understand this but then I don't really.

**Comment by fauust** [comment]
As far as I understand https://github.com/MariaDB/server/commit/b9fee60871264ea56a5582ca7a436aabdabea6fc, because of the test suite. But other than that, I really don't know.

**Comment by fauust** [comment]
Maybe I did not respond to your question @illuusio? Is it about `Recommends` vs `Suggests`?
I had proposed to use `Suggest` at the begining but there is maybe a good reason why this is not sufficient? That's why looking at what we do with RPM could maybe help us to decide? 

[Here](https://www.debian.org/doc/debian-policy/ch-relationships.html#binary-dependencies-depends-recommends-suggests-enhances-pre-depends
) is what Debian policy says about the topic:

> Recommends
>
>    This declares a strong, but not absolute, dependency.
>    The Recommends field should list packages that would be found together with this one in all but unusual installations.
>
> Suggests
>
>    This is used to declare that one package may be more useful with one or more others. Using this field tells the packaging system and the user that the listed packages are related to this one and can perhaps enhance its usefulness, but that installing this one without them is perfectly reasonable.

**Comment by ottok** [comment]
Good reference to the policy @fauust. I am actually not sure should we go with `Suggests` or `Recommends`. Seems `Suggests` might be the looser and better option after all?

**Comment by fauust** [comment]
Yep, installing MariaDB without Galera seems "perfectly reasonable." to me.

**Comment by illuusio** [comment]
@fauust Actually I was thinking is MariaDB build against Galera or is just optional and additional  software that can be used without compiling it 'in'. If it's just additional software then  Suggests should be correct. If it's needed as build dependency then it can't be on suggests but hard dependency.

**Comment by grooverdan** [comment]
Details on galera RPM dependencies are in cmake/cpack_rpm.cmake

**Comment by illuusio** [comment]
@grooverdan and @fauust I tooks a look how RPM is made and if it's wanted as DEB there should be  CPACK_RPM_server_PACKAGE_SUGGESTS in WITH_WSREP if in file  cmake/cpack_rpm.cmake where only thing suggested is galera-4 like:
```
SETA(CPACK_RPM_server_PACKAGE_REQUIRES
    "rsync" "lsof" "grep" "gawk" "iproute"
    "coreutils" "findutils" "tar")

SETA(CPACK_RPM_server_PACKAGE_SUGGESTS
    "galera-4")
```
 But this problematic as RHEL does not support spec/RPM **Suggests:** but it should just ignore it. I can also make a PR to this PR if needed.

**Comment by illuusio** [comment]
@fauust RPM suggests are kind of 'new' feature (https://fedoraproject.org/wiki/Packaging:WeakDependencies) and it can be turned of in DNF (https://bgstack15.wordpress.com/2017/05/13/dnf-ignore-weak-dependencies/) and in SUSE world in zypper there is recommends and suggests that shows packages which are needed but I think we just trust that user have knowledge to install galera if it's needed.
It can be on package descriptions as it's kind of new in RPM world and if it's rising eye browns then just add it aftwerwards?

**Comment by illuusio** [comment]
I'll test building this and packages are there any drawback but I can't see any

**Comment by vuvova** [comment]
We don't use `Suggests:` but we do use `Recommends:`.
cpack handles it and converts `Recommends:` to `Requires:` if rpm is too old

**Comment by illuusio** [comment]
@vuvova ok is there cmake some way to make 

> We don't use `Suggests:` but we do use `Recommends:`. cpack handles it and converts `Recommends:` to `Requires:` if rpm is too old

@vuvova that is good policy. There is no `Recommends:` CMake in basic macros but if I undestood correct there is internal one in _cmake/Internal/CPack/CPackRPM.cmake_? I edited my previous suggest comment to fit in and @fauust  then we should add this recommends to rpm to be in line with DEB?

**Comment by grooverdan** [comment]
@illuusio added. The other adjacent package dependencies are those used with galera to perform the SST mechanism. So they are definitely not Depends any more. Their status in Recommends does look like its an independent choice from Galera.

@janlindstrom do you think it makes sense for the SST Depend(ant) packages to be attached as Depends in the galera packaging directly?

**Comment by illuusio** [comment]
> @illuusio added. The other adjacent package dependencies are those used with galera to perform the SST mechanism. So they are definitely not Depends any more. Their status in Recommends does look like its an independent choice from Galera.

@grooverdan sorry I didn't right get this? SST mechanism? So they should be Suggests or removed. Should they be own metapackage?

**Comment by dr-m** [comment]
@illuusio Galera snapshot transfer (SST) is implemented in shell scripts, such as `scripts/wsrep_sst_common.sh`. It could make sense to create a separate package for those scripts that would be suggested by the MariaDB server package, and make that suggested package `Depends:` on Galera and the other dependencies.

While the server can be compiled without any reference to Galera (`cmake -DWITH_WSREP=OFF`), the default is to build a server that may be used with the Galera library. Before MariaDB Server 10.1, there were separate Galera-aware packages.

**Comment by grooverdan** [comment]
Also scripts `galera_recovery` and `galera_new_cluster`

**Comment by illuusio** [comment]
> While the server can be compiled without any reference to Galera (`cmake -DWITH_WSREP=OFF`), the default is to build a server that may be used with the Galera library. Before MariaDB Server 10.1, there were separate Galera-aware packages.

I don't see compiling with Galera on and having external package that has these scripts inside. I'm just again thinking is this PR for all this or just remove galera from dependencies and make new PR to add this new package?

**Comment by grooverdan** [comment]
@vuvova are you ok with a new package again? mariadb-server-galera-10.[78] which picks up the sst + `galera_*` systemd scripts to a new package and let that package pick up the dependencies of galera + other things which are needed by the sst scripts.

**Comment by ottok** [comment]
Why over-engineer it with a new package?

That is going to require a lot of work to actually make it a smooth
install/upgrade experience. The PR already looks good and makes Galera a
weaker dependency, but it is still a dependency and a couple of extra
scripts in the server do not matter, and in fact are quite justified as
long as the server is also compiled with Galera support.

**Comment by grooverdan** [comment]
@ottok you're missing the point. If galera and the sst script dependencies are optional, installing `galera-4` as the wsrep provider without the lsof/rsync/socat will result in a system without a functioning SST (which is a core component of a working Galera). The server will still be compiled with Galera support.

If you are ok with with the following as server `Depends`  `gawk, iproute2 [linux-any], lsof [linux-any], rsync,  socat` without any real purpose (until galera-4 is installed) that's ok I suppose too.

**Comment by dr-m** [comment]
>Why over-engineer it with a new package?

@ottok I believe that a new package for the Galera specific dependencies is technically the cleanest solution that will both avoid "polluting" the main package with unnecessary dependencies, and guarantee that whoever wants to use Galera with MariaDB will get all the required dependencies installed easily.

If creating a new package really requires major effort, perhaps it could be acceptable to make the simpler change as a stopgap measure. But then again, temporary solutions too easily become permanent ones.

**Comment by vuvova** [comment]
is that urgent? can it wait for Robert to finish https://jira.mariadb.org/browse/MDEV-7502?
That work might use some of SST scripts in the server without galera

**Comment by ottok** [comment]
> @ottok you're missing the point. If galera and the sst script dependencies are optional, installing `galera-4` as the wsrep provider without the lsof/rsync/socat will result in a system without a functioning SST (which is a core component of a working Galera). The server will still be compiled with Galera support.
> 
> If you are ok with with the following as server `Depends` `gawk, iproute2 [linux-any], lsof [linux-any], rsync, socat` without any real purpose (until galera-4 is installed) that's ok I suppose too.

We can add those packages as dependencies of `galera-4`, so that `apt install mariadb-server galera-4` pulls in all that is needed, yet running `apt install mariadb-server` alone will not pull in `galera-4` or it's (future) dependencies such as `gawk, iproute2 [linux-any], lsof [linux-any], rsync, socat`.

**Comment by illuusio** [comment]
> We can add those packages as dependencies of `galera-4`, so that `apt install mariadb-server galera-4` pulls in all that is needed, yet running `apt install mariadb-server` alone will not pull in `galera-4` or it's (future) dependencies such as `gawk, iproute2 [linux-any], lsof [linux-any], rsync, socat`.

I checked galera-4 package and it does not need those dependencies. As it would be easiest and most efficient script which I understood that are Galera specific need these are in MariaDB server git:

 * scripts/wsrep_sst_mysqldump.sh needs awk
 * scripts/wsrep_sst_rsync.sh needs rsync
 * scripts/wsrep_sst_mariabackup.sh needs socat
 * scripts/wsrep_sst_common.sh needs iproute2
 * 
What happens if these are weak dependencies on MariaDB server and also in galera-4 does they get installed when you install galera-4 with MariaDB server? I don't know if apt is that clever?

**Comment by illuusio** [comment]
I opened JIRA bug for this: [MDEV-28284](https://jira.mariadb.org/browse/MDEV-28284) to summary that:
 * **libdbi-perl** can be moved to `mariadb-test` package dependency and removed altogether from `mariadb-server`
 * **socat** can be optional if we don't have Galera 4 as it's only used in there but should moved hard dependency of `mariadb-test` as it required there
 * **(g)awk** can't be optional as `mysql_install_db.sh` is requiring it and it's used in Debian `mariadb-server` preinst-script
 * **iproute2** is used in several places 
 * **rsync** can be optional for `mariadb-server` but is requires by `mysql-test` and `mariadb-backup`. So if it's optional here then it must be hard dependency there
  * **lsof** can't be optional as it's used `mariadb-server` postinst-script

**Comment by illuusio** [comment]
Build with changes that I suggested can be found on: https://salsa.debian.org/illuusio/mariadb-server/-/pipelines/370768

**Comment by fauust** [comment]
@illuusio let me know if that looks OK to you now.

**Comment by illuusio** [comment]
> @illuusio let me know if that looks OK to you now.

RPM seems to need handled same way and needs some care and loving because dependencies are not correct according to my study.
Update RPM same Required and same recommends. Otherwise starting to be okay with me.

**Comment by ottok** [comment]
> I opened JIRA bug for this: [MDEV-28284](https://jira.mariadb.org/browse/MDEV-28284) to summary that:
> 
> * **libdbi-perl** can be moved to `mariadb-test` package dependency and removed altogether from `mariadb-server`
> * **socat** can be optional if we don't have Galera 4 as it's only used in there but should moved hard dependency of `mariadb-test` as it required there
> * **(g)awk** can't be optional as `mysql_install_db.sh` is requiring it and it's used in Debian `mariadb-server` preinst-script
> * **iproute2** is used in several places
> * **rsync** can be optional for `mariadb-server` but is requires by `mysql-test` and `mariadb-backup`. So if it's optional here then it must be hard dependency there
> * **lsof** can't be optional as it's used `mariadb-server` postinst-script

Great summary @illuusio! I reviewed the Debian commit https://github.com/MariaDB/server/pull/1829/commits/7d03d9042fcc517854efc3393a1e6d47e32dbe15 by @fauust and it does not seem to match this. For example it still moves 'lsof` from Depends to Recommends which was against what Tuukka wrote above.

I also checked pipeline https://salsa.debian.org/illuusio/mariadb-server/-/pipelines/370768 of branch https://salsa.debian.org/illuusio/mariadb-server/-/commits/10.7-MDEV-28284-remove-galera-4 with commit https://salsa.debian.org/illuusio/mariadb-server/-/commit/582f59c29be9d24f16942e6daf76f894144b91fb. That looks better but that commit ID is not on this PR.

**Comment by ottok** [comment]
Related, do you want me to add some new `Depends` in https://salsa.debian.org/mariadb-team/galera-4/-/blob/master/debian/control? Or if you submit a MR there, I will be happy to review and merge.

**Comment by illuusio** [comment]
I'll try to make MR for the Galera as it's only viable option.

File: cmake/cpack_rpm.cmake [hunk 1]
```diff
@@ -223,7 +223,7 @@ SETA(CPACK_RPM_server_PACKAGE_REQUIRES
   "MariaDB-client >= 10.6.1")
 
 IF(WITH_WSREP)
-  SETA(CPACK_RPM_server_PACKAGE_REQUIRES
+  SETA(CPACK_RPM_server_PACKAGE_RECOMMENDS
```
Comment (SergMariaDB) [inline cmake/cpack_rpm.cmake:226]: @grooverdan, `"rsync" "lsof" "grep" "gawk" "iproute"
     "coreutils" "findutils" "tar"`
seem must stay in CPACK_RPM_server_PACKAGE_REQUIRES, right? 
Please, consider adding new line for CPACK_RPM_server_PACKAGE_RECOMMENDS.

Comment (illuusio) [reply cmake/cpack_rpm.cmake:226]: I think so too as they are needed several stuff in scripts.

**Comment by ottok** [comment]
Galera is still `Depends` in MariaDB:
```
$ grep galera upstream-10.11-debian/control upstream-11.0-debian/control mariadb-server/debian/control 
upstream-10.11-debian/control:Depends: galera-4 (>= 26.4),
upstream-11.0-debian/control:Depends: galera-4 (>= 26.4),
mariadb-server/debian/control:Depends: galera-4 (>= 26.4),
```

If/when made into `Recommends`, we need to remember to also extend `salsa-ci.yml` to specifically install Galera and test it so that the weaker dependency does not result with in Galera being omitted from install/upgrade tests.

**Comment by illuusio** [comment]
Should be have separate MariaDB package for `galera so we could move all these `wsrep_` stuff to it and have correct dependencies?

**Comment by ottok** [comment]
Yes, we could have package mariadb-plugin-galera starting from 11.0 if Serg
thinks it makes sense.

**Comment by grooverdan** [comment]
or:
```
Package: mariadb-server-galera
...
Depends: mariadb-server,
                   galera-4,
                   rsync, lsof, grep, gawk, iproute, coreutils, findutils, tar
```

I don't mind the look of `mariadb-plugin-galera` however (once I get past the grimaces about if galera is really a plugin being so embedded in to the server).

**Comment by iangilfillan** [comment]
[MDEV-28284](https://jira.mariadb.org/browse/MDEV-28284) was closed as Won't Do, can this PR be closed as well?

**Comment by grooverdan** [comment]
continued in #4644.
